### PR TITLE
DOC: Fix creation of [source] links in the doc creation

### PR DESCRIPTION
**Pull request by vkk800**
See this issue: https://github.com/pandas-dev/pandas/issues/23046

A simple fix is to add `inspect.unwrap(obj)` before `getsourcefile`. This follows the `__wrapped__` attributes of the functions "all the way to the bottom" and then returns the object found.

One problem with this is that `unwrap` function was only introduced in Python 3.4. This means that compiling the documentation will now fail on Python versions older than that. I do not know if this is acceptable.

### CODE REVIEW

**Comment by datapythonista** [comment]
In the CI we're building the documentation in Python 3.6, and in couple of months we'll only support Python 3.5+, so I think it's acceptable.

Could you test the broken links locally?

Btw, unrelated to this PR, in `scripts/validate_docstrings.py` we've got some tricky things to solve the same problem. May be you can take a look and see if they can be simplified with `inspect.unwrap`? We'are probably reinventing the wheel there.

**Comment by datapythonista** [comment]
@vkk800 there is no problem with the script, it's just that it's possible that we reimplemented `inspect.unwrap` here: https://github.com/pandas-dev/pandas/blob/master/scripts/validate_docstrings.py#L169

As you already worked with that function, I thought it'd be useful if you can take a look, see if that's the case, and if we can delete that method and simply use `inspect.unwrap`, a PR would be great too. No issue, but feel free to create one.

**Comment by datapythonista** [comment]
@jorisvandenbossche I think you mmay want to take a look at this PR

**Comment by jorisvandenbossche** [comment]
For now, we can still call the `unwrap` conditionally on the python version?

**Comment by vkk800** [comment]
> For now, we can still call the `unwrap` conditionally on the python version?

We could do something like `if unwrap in dir(inspect)`, but this is a bit ugly (and confusing years from now when no-one remembers why it was originally added). If no-one is compiling the docs with Python older than 3.4, I don't really see the point.

I took a look at the `validate_docstrings` and it looks that `_to_original_callable` could probably be improved by adding `unwrap`, but not removed completely, because `unwrap()` still wouldn't find the functions hidden behind `functools.partial` or class methods as `_to_original_callable` is trying to do. Also, the only thing that `unwrap()` achieves better than using `@functools.wraps` decorator is finding the original source file; docstrings, etc. are handled just fine with `@functools.wraps`.

**Comment by jreback** [comment]
@datapythonista

**Comment by datapythonista** [comment]
@vkk800 if you can simply add a `if sys.version_info >= (3, 5):` that would be great. For such a small change let's keep the doc building working on Python 2.

**Review by datapythonista — APPROVED** [review]
lgtm, thanks @vkk800

File: doc/source/conf.py [hunk 1]
```diff
@@ -569,7 +569,11 @@ def linkcode_resolve(domain, info):
             return None
 
     try:
-        fn = inspect.getsourcefile(obj)
+        # inspect.unwrap() was added in Python version 3.4
```
Comment (jreback) [inline doc/source/conf.py]: we don't actually support 3.4, make this 3.5

File: doc/source/conf.py [hunk 2]
```diff
@@ -569,7 +569,11 @@ def linkcode_resolve(domain, info):
...
+        if sys.version_info >= (3, 4):
```
Comment (datapythonista) [inline doc/source/conf.py]: For a review like this I think it's useful the new github feature to suggest changes (the icon with the +- icon when adding a comment. This way, afaik, the user can make the change with a single click.

```suggestion
        if sys.version_info >= (3, 5):
```

**Comment by jreback** [comment]
thanks @vkk800

**Pull request merged**